GooglePlayInstant/LoadingScreen/LoadingScreen.cs

```diff
@@ -0,0 +1,121 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections;
+using UnityEngine;
+using UnityEngine.Networking;
+using UnityEngine.SceneManagement;
+using UnityEngine.UI;
+
+namespace GooglePlayInstant.LoadingScreen
+{
+    /// <summary>
+    /// Downloads the AssetBundle available at AssetBundleUrl and updates LoadingBar with its progress
+    /// </summary>
+    public class LoadingScreen : MonoBehaviour
+    {
+        [Tooltip("The url used to fetch the AssetBundle on Start")]
+        public string AssetBundleUrl;
+
+        [Tooltip("The LoadingBar used to indicated download and install progress")]
+        public LoadingBar LoadingBar;
+
+        public RawImage Background;
+
+        private const int MaxAttemptCount = 3;
+        private AssetBundle _bundle;
+        private int _assetBundleRetrievalAttemptCount;
+
+        private IEnumerator Start()
+        {
+            while (_bundle == null && _assetBundleRetrievalAttemptCount < MaxAttemptCount)
+            {
+                yield return GetAssetBundle(AssetBundleUrl);
+            }
+
+            if (_bundle == null)
+            {
+                Debug.LogError("AssetBundle failed to be downloaded.");
+                yield break;
+            }
+
+            var sceneLoadOperation = SceneManager.LoadSceneAsync(_bundle.GetAllScenePaths()[0]);
+            yield return LoadingBar.FillUntilDone(sceneLoadOperation, LoadingBar.AssetBundleDownloadMaxProportion, 1f);
+        }
+
+        private IEnumerator GetAssetBundle(string assetBundleUrl)
+        {
+            UnityWebRequest webRequest;
+            var downloadOperation = StartAssetBundleDownload(assetBundleUrl, out webRequest);
+
+            yield return LoadingBar.FillUntilDone(downloadOperation,
+                0f, LoadingBar.AssetBundleDownloadMaxProportion);
+
+            if (DidRequestFail(webRequest))
+            {
+                yield return HandleRequestFailed(assetBundleUrl, webRequest);
+            }
+            else
+            {
+                _bundle = DownloadHandlerAssetBundle.GetContent(webRequest);
+            }
+        }
+
+        private IEnumerator HandleRequestFailed(string assetBundleUrl, UnityWebRequest webRequest)
+        {
+            if (_assetBundleRetrievalAttemptCount < MaxAttemptCount)
+            {
+                _assetBundleRetrievalAttemptCount++;
+                Debug.LogFormat("Attempt #{0} at downloading AssetBundle...", _assetBundleRetrievalAttemptCount);
+                yield return new WaitForSeconds(2);
+
+                //TODO: revisit this methodology of setting the loading bar
+                LoadingBar.SetProgress(0f);
+
+                yield return GetAssetBundle(assetBundleUrl);
+            }
+            else
+            {
+                Debug.LogErrorFormat("Error downloading asset bundle: {0}", webRequest.error);
```

Comment:
Is it worth putting `yield break;` here?

Reply:
I sort of see the value in it, in case anyone adds code to the end of the function assuming it'll only run if the download completes. However, JetBrains greys it out as a "redundant control flow jump statement" making it likely that it'll just get deleted in the future.

Reply:
Sounds good. No changes here then.